sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs

```diff
@@ -0,0 +1,238 @@
+                {
+                    ExceptionDispatchInfo.Capture(thisTryException).Throw();
+                }
+
+                throw new AggregateException($"Retry failed after {message.RetryCount + 1} tries.", allTryExceptions);
+            }
+
+            // ShouldRetry returned false this iteration and the last request
+            // we sent didn't cause an exception to be thrown.
+            // So, we're done.  Exit the while loop.
+            break;
+        }
+    }
+
+    protected virtual void OnSendingRequest(PipelineMessage message) { }
+
+    protected virtual ValueTask OnSendingRequestAsync(PipelineMessage message) => default;
+
+    protected virtual void OnRequestSent(PipelineMessage message) { }
+
+    protected virtual ValueTask OnRequestSentAsync(PipelineMessage message) => default;
+
+    protected virtual void OnTryComplete(PipelineMessage message) { }
+
+    internal bool ShouldRetry(PipelineMessage message, Exception? exception)
+    {
+        // If there was no exception and we got a success response, don't retry.
+        if (exception is null && message.Response is not null && !message.Response.IsError)
+        {
+            return false;
+        }
+
+        return ShouldRetryCore(message, exception);
+    }
+
+    internal async ValueTask<bool> ShouldRetryAsync(PipelineMessage message, Exception? exception)
+    {
+        // If there was no exception and we got a success response, don't retry.
+        if (exception is null && message.Response is not null && !message.Response.IsError)
+        {
+            return false;
+        }
+
+        return await ShouldRetryCoreAsync(message, exception).ConfigureAwait(false);
+    }
+
+    protected virtual bool ShouldRetryCore(PipelineMessage message, Exception? exception)
+    {
+        if (message.RetryCount >= _maxRetries)
+        {
+            // We've exceeded the maximum number of retries, so don't retry.
+            return false;
+        }
+
+        return exception is null ?
+            IsRetriable(message) :
+            IsRetriable(message, exception);
+    }
+
+    protected virtual ValueTask<bool> ShouldRetryCoreAsync(PipelineMessage message, Exception? exception)
+        => new(ShouldRetryCore(message, exception));
+
+    internal TimeSpan GetNextDelay(PipelineMessage message, int tryCount)
+    => GetNextDelayCore(message, tryCount);
+
+    protected virtual TimeSpan GetNextDelayCore(PipelineMessage message, int tryCount)
+    {
+        // Default implementation is exponential backoff
+        return TimeSpan.FromMilliseconds((1 << (tryCount - 1)) * _initialDelay.TotalMilliseconds);
+    }
+
+    protected virtual async Task WaitAsync(TimeSpan time, CancellationToken cancellationToken)
+    {
+        await Task.Delay(time, cancellationToken).ConfigureAwait(false);
+    }
+
+    protected virtual void Wait(TimeSpan time, CancellationToken cancellationToken)
+    {
+        cancellationToken.WaitHandle.WaitOne(time);
```

Comment:
Interesting!  I think I had just copied this in from Azure.Core, but I can look at it again to see if I moved it over incorrectly or if it's an existing issue.

Reply:
It looks like this is [the same implementation we have in Azure.Core today](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/Pipeline/RetryPolicy.cs#L173-L176) ... @christothes, do you have a suggestion for how we would want to modify this code?  Any risk that we would start throwing a new exception where we weren't before, or is it just that we would be correctly handling a CancellationToken by changing this, where we had a bug before?

Reply:
Fixing this would match the behavior of the async path, I think. My suggestion would be something like:

```c#
if (cancellationToken.WaitHandle.WaitOne(time))
{
    throw new TaskCanceledException();
}
```

We must be missing test coverage on this also, so it would be nice to validate with some tests.

Reply:
We should use [CancellationHelper](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/System.ClientModel/src/Internal/CancellationHelper.cs) to make sure we are consistent with how these are thrown.

Reply:
We still need to check if the token was cancelled. WaitOne will return true even if the token wasn't signaled.

Reply:
Addressed in ClientModel and filed https://github.com/Azure/azure-sdk-for-net/issues/41400 to track the outstanding issue in Azure.Core.